### ISSUE TITLE
[simclasses] I3Photon is entirely too large (Trac #1978)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#1970]
`sizeof(I3Photon)` is 368 bytes, more than twice (!) the size of I3Particle. This becomes problematic at high energies when can you have tens of millions of them that need to be stored in the frame between I3CLSimModule and I3PhotonToMCPEConverter. There are a couple of ways to compact things:
- Remove all fields except final position, direction, time, wavelength, weight, and group velocity. These are strictly needed to produce PEs; everything else is convenience and fluff.
- Store time, wavelength, weight, and group velocity as floats. This is the storage they use in I3CLSimPhoton, so this just delays any conversion to double precision to I3PhotonToMCPEConverter's stack. No information is lost.
- Likewise, store x,y,z,theta,phi as floats, as in I3CLSimPhoton. `I3Position` and `I3Direction` are ridiculously large (72 and 56 bytes, respectively).

There is still a case to be made for a general-purpose photon object, and the history and source-tracking fields are potentially useful in special situations. In light of that, I propose a new object, `I3BasicPhoton`, with more or less this layout:

```text
class I3BasicPhoton {
    float x, y, z;
    float theta, phi;
    float time;
    float wavelength;
    float groupvel;
    float weight;
};
```
that is 36 bytes, or 10x smaller than I3Photon. I3CLSimModule could emit these instead of I3Photons in some as-yet-to-be-implemented memory efficiency mode.

Migrated from https://code.icecube.wisc.edu/ticket/1978
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:13:52", 
    "description": "`sizeof(I3Photon)` is 368 bytes, more than twice (!) the size of I3Particle. This becomes problematic at high energies when can you have tens of millions of them that need to be stored in the frame between I3CLSimModule and I3PhotonToMCPEConverter. There are a couple of ways to compact things:\n- Remove all fields except final position, direction, time, wavelength, weight, and group velocity. These are strictly needed to produce PEs; everything else is convenience and fluff.\n- Store time, wavelength, weight, and group velocity as floats. This is the storage they use in I3CLSimPhoton, so this just delays any conversion to double precision to I3PhotonToMCPEConverter's stack. No information is lost.\n- Likewise, store x,y,z,theta,phi as floats, as in I3CLSimPhoton. `I3Position` and `I3Direction` are ridiculously large (72 and 56 bytes, respectively).\n\nThere is still a case to be made for a general-purpose photon object, and the history and source-tracking fields are potentially useful in special situations. In light of that, I propose a new object, `I3BasicPhoton`, with more or less this layout:\n\n{{{\nclass I3BasicPhoton {\n    float x, y, z;\n    float theta, phi;\n    float time;\n    float wavelength;\n    float groupvel;\n    float weight;\n};\n}}}\nthat is 36 bytes, or 10x smaller than I3Photon. I3CLSimModule could emit these instead of I3Photons in some as-yet-to-be-implemented memory efficiency mode.\n", 
    "reporter": "jvansanten", 
    "cc": "claudio.kopper", 
    "resolution": "fixed", 
    "_ts": "1550067232264679", 
    "component": "combo simulation", 
    "summary": "[simclasses] I3Photon is entirely too large", 
    "priority": "normal", 
    "keywords": "", 
    "time": "2017-04-01T18:23:27", 
    "milestone": "", 
    "owner": "jvansanten", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20170401T18:25:21: claudio.kopper commented:

> Note that `float time` might not be enough if we decide to simulate very long events. Otherwise I think this is a good idea.

**Comment by icecube-trac** [comment]
Trac update at 20170401T18:31:31: jvansanten commented:

> That's clear from our past experience of dabbling in single precision. To discourage potentially dangerous use, I would probably only GetX() methods that return double/I3Position/I3Direction, but no corresponding setters. 
>

**Comment by icecube-trac** [comment]
Trac update at 20170403T13:49:41: jvansanten commented:

> In changeset 154595:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="154595"
> Slim I3CompressedPhoton further.
> 
> Current size is 56 bytes, including 4 bytes of padding required to align
> the double and int64 fields. Storing the time in single precision (which
> is potentially dangerous as noted in #1970), would reduce it to 48 bytes,
> and omitting the source particle IDs would bring it down to an absolute
> minimum of 36 bytes. (See #1970)
> ```

**Comment by icecube-trac** [comment]
Trac update at 20170403T14:11:24: 
* jvansanten commented:

> In changeset 154598:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="154598"
> Emit I3CompressedPhotons if PhotonHistoryEntries == 0.
> 
> This reduces the part of I3CLSimModule's memory footprint that is proportional
> to deposited energy by a factor of 6. The same applies to the output file
> size, if photons are being stored. This could be increased to 10 if we were
> willing to store the photon time in single precision (implies short events)
> and omit the source particle IDs. Closes #1970.
> ```


* jvansanten changed resolution from "" to "fixed"
* jvansanten changed owner from "" to "jvansanten"
* jvansanten changed status from "new" to "closed"

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:13:52: 
* olivas changed milestone from "April 2017" to ""
* olivas commented:

> Milestone April 2017 deleted
